app/cdap/components/Metadata/ComplexSchema/index.tsx

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import React, { useEffect, useReducer } from 'react';
+import T from 'i18n-react';
+import styled from 'styled-components';
+import { SCHEMA_TYPES, getSchemaState } from 'components/Metadata/ComplexSchema/helper';
+import Checkbox from '@material-ui/core/Checkbox';
+import Input from '@material-ui/core/Input';
+import Select from '@material-ui/core/Select';
+import MenuItem from '@material-ui/core/MenuItem';
+
+const I18N_PREFIX = 'features.MetadataSummary';
+
+const SchemaTable = styled.table`
+  width: 100%;
+  tr > th {
+    padding: 6px 0;
+  }
+  tr > td:first-child,
+  tr > th:first-child {
+    padding-left: 10px;
+    padding-right: 10px;
+  }
+  tr > td:nth-child(2n) {
+    width: 90px;
+  }
+  tr > td:nth-child(3n) {
+    width: 50px;
+  }
+`;
+
+const FieldRow = styled.tr`
+  box-shadow: 2px 0 2px 0 rgb(1 0 0 / 20%);
+  ${({ hasError }) => hasError && `border: solid 2px var(--danger);`}
+  ${({ isNotNested }) => isNotNested && `box-shadow: none; border-top: solid 1px var(--grey06);`}
+`;
+
+const FieldInput = styled(Input)`
+  width: 100%;
+  color: var(--grey02) !important;
+  :before {
+    border: none !important;
+  }
+`;
+
+const FieldSelect = styled(Select)`
+  width: 100%;
+  color: var(--grey02) !important;
+  :before {
+    border: none !important;
+  }
+`;
+
+const ErrorText = styled.td`
+  color: var(--danger);
+`;
+
+interface ISchema {
+  type: string;
+  name: string;
+  fields: { [key: string]: string };
+  getFields: () => {};
+}
+
+interface IInputProps {
+  recordName: string;
+  isRecordSchema: boolean;
+  typeIndex: string;
+  parentFormatOutput: () => void;
+  isDisabled: boolean;
+  schemaPrefix: string;
+  derivedDatasetId: string;
+  isInputSchema: boolean;
+  isInStudio: boolean;
+  errors: { [key: string]: string };
+}
+
+interface IComplexSchemaProps {
+  schema: ISchema;
+  inputProps: IInputProps;
+}
+
+const ComplexSchema: React.FC<IComplexSchemaProps> = ({ schema, inputProps }) => {
+  const initialState = {
+    parsedSchema: [],
+    emptySchema: false,
+    error: '',
+    disabledTooltip: null,
+    updateDefault: false,
+    model: {},
+  };
+  function reducer(state, action) {
```

Comment:
At this point, it looks like the reducer isn't doing anything. We should remove the reducer or refactor `getSchemaStore` to be action/reducer based.

Reply:
Since it's a parsing logic moved it a helper method and useState instead of reducer.